### PR TITLE
JUnit bump 4 -> 5

### PR DESCRIPTION
**Pull request by bvfalcon**
Migrating unit-tests from JUnit 4 to JUnit 5 ([Issue](https://github.com/eclipse-ee4j/jersey/issues/4902)).

Notes:
1) **IMPORTANT**: Annotations `@Before` and `@After` of JUnit 4 work a little another way, as `@BeforeEach` and `@AfterEach` of JUnit 5. In case of overriding annotations `@Before` and `@After` will work, annotations `@BeforeEach` and `@AfterEach` - not. See [example](https://github.com/bvfalcon/junit-4-5-difference-test) for demonstration.
2) [jersey-release-notes-maven-plugin](https://github.com/eclipse-ee4j/jersey/tree/master/tools/jersey-release-notes-maven-plugin) still uses JUnit 4 because of used in unit-tests maven-plugin-testing-harness has no support for JUnit5. And because of it is comparatively isolated part of jersey project.
3) [jersey-tests-osgi-functional](https://github.com/eclipse-ee4j/jersey/tree/master/tests/osgi/functional), [osgi-helloworld-webapp](https://github.com/eclipse-ee4j/jersey/tree/master/examples/osgi-helloworld-webapp), [osgi-http-service](https://github.com/eclipse-ee4j/jersey/tree/master/examples/osgi-http-service) and [extended-wadl-webapp](https://github.com/eclipse-ee4j/jersey/tree/master/examples/extended-wadl-webapp) still use JUnit 4 because of pax-exam has no support for JUnit 5 (issues [one](https://github.com/ops4j/org.ops4j.pax.exam2/issues/886), [two](https://github.com/ops4j/org.ops4j.pax.exam2/issues/951))
4) I suppose, test [monitoring](https://github.com/eclipse-ee4j/jersey/tree/master/tests/performance/test-cases/monitoring) never worked: [dependencies without version](https://github.com/eclipse-ee4j/jersey/blob/c81fc8fb23202ab8fc08d92630ea9e06bcf188d6/tests/performance/test-cases/monitoring/pom.xml#L68-L71) cannot work in projects without parent.

### CODE REVIEW

**Comment by jansupol** [comment]
@bvfalcon Thank you for doing this. It is completely fine to keep some tests on Junit 4.

The only requirement is that Jersey test-framework should keep supporting JUnit 4 as well as Junit 5, which I suppose it does.

Your PR is very huge. I'd suggest to finish the PR with what you have (so that the tests pass) and continue with migrating the remaining tests (if any) in another PR.

**Comment by bvfalcon** [comment]
Ready for review.
Some notices about last builds and their errors:

1) AccessControl access denied  errors on CI Jenkins JDK 19

```
[ERROR] Errors: 
[ERROR]   ByteBufferInputStreamTest.testBlockingReadByteArray:408 » AccessControl access denied ("java.lang.RuntimePermission" "setContextClassLoader")
[ERROR]   ByteBufferInputStreamTest.testBlockingReadSingleByte:340 » AccessControl access denied ("java.lang.RuntimePermission" "setContextClassLoader")
```

JDK 19 -- it seems to me, configration (namely SecurityManager) is incorrect. This is the one explains errors (for example [full log](https://ci.eclipse.org/jersey/job/continuous-integration/job/PR-5123/41/execution/node/61/log/) of last build on JDK 19)

2) Error in test of [jersey-4697](https://github.com/eclipse-ee4j/jersey/blob/c81fc8fb23202ab8fc08d92630ea9e06bcf188d6/tests/integration/jersey-4697/src/test/java/org/glassfish/jersey/tests/integration/jersey4697/MonitoringEventListenerTest.java#L124-L142)

```
[ERROR] Failures: 
[ERROR]   MonitoringEventListenerTest.exceptionInScheduler:140->queueIsEmpty:153 expected: <0> but was: <2>
```

Test looks like a fragile (depends on timeouts). I see this errors in other PR and think, that it was not caused by changes in this my PR.

**Comment by jansupol** [comment]
There is no Security Manager on JDK 19. Some SecurityManager related tests might fail there (The tests that set some permission and check the permission is set). But, the build on JDK 19 does pass in [this PR](https://github.com/eclipse-ee4j/jersey/pull/5124).

**Comment by jansupol** [comment]
`MonitoringEventListenerTest.exceptionInScheduler:140->queueIsEmpty:153 expected: <0> but was: <2>` Intermittently fails, especially on some OSs

**Comment by jansupol** [comment]
@bvfalcon Sorry it takes me long. It is a large PR.

File: ext/wadl-doclet/src/main/java12/org/glassfish/jersey/wadl/doclet/ResourceDoclet.java [hunk 1]
```diff
@@ -190,7 +190,6 @@ public boolean run(DocletEnvironment docEnv) {
                     }
                     result.getDocs().add(classDocType);
                     success = DocletUtils.createOutputFile(optionOutput.getValue(), docProcessor, result);
-
```
Comment (senivam) [inline ext/wadl-doclet/src/main/java12/org/glassfish/jersey/wadl/doclet/ResourceDoclet.java:193]: why?

File: media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/annotation/JacksonFeatures.java [hunk 2]
```diff
@@ -26,7 +26,7 @@
      * Deserialization features to disable.
      */
     public DeserializationFeature[] deserializationDisable() default { };
-    
+
```
Comment (senivam) [inline media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/annotation/JacksonFeatures.java:29]: why?

File: media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/EndpointConfigBase.java [hunk 3]
```diff
@@ -53,7 +53,7 @@
     protected EndpointConfigBase(MapperConfig<?> config) {
         _config = config;
     }
-    
+
```
Comment (senivam) [inline media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/EndpointConfigBase.java:56]: I think formatting is not a purpose of the PR, so I would recommend not doing unnecessary formatting in files where there are no other changes.

File: media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/JaxRSFeature.java [hunk 4]
```diff
@@ -32,7 +32,7 @@ public enum JaxRSFeature implements ConfigFeature
     /* HTTP headers
     /**********************************************************
      */
-    
+
```
Comment (senivam) [inline media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/JaxRSFeature.java:35]: same note regarding unnecessary formatting

**Review by senivam — CHANGES_REQUESTED** [review]
Please kindly remove all unnecessary formatting from the PR - there is a bunch of files that are modified only by empty lines or spaces and this increases the number of files to be reviewed and introduces unnecessary changes to the project. I will proceed with the review after all redundant formatting is cleared.

File: ext/wadl-doclet/src/main/java12/org/glassfish/jersey/wadl/doclet/ResourceDoclet.java (hunk 1 above)
Comment (bvfalcon) [reply ext/wadl-doclet/src/main/java12/org/glassfish/jersey/wadl/doclet/ResourceDoclet.java:193]: You are right: this was not necessary. [Checkstyle rule](https://github.com/eclipse-ee4j/jersey/blob/master/etc/config/checkstyle-verify.xml#L132-L136) was incorrect understood by me. Solved here and in all similar occurencies

File: media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/EndpointConfigBase.java (hunk 3 above)
Comment (bvfalcon) [reply media/json-jackson/src/main/java/org/glassfish/jersey/jackson/internal/jackson/jaxrs/cfg/EndpointConfigBase.java:56]: Yes,as I have already say, it was incorrect understanding of checkstyle rules.

File: archetypes/jersey-quickstart-grizzly2/src/main/resources/archetype-resources/pom.xml [hunk 5]
```diff
@@ -38,9 +38,9 @@
         </dependency>
         -->
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <version>4.12</version>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>5.9.0</version>
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply archetypes/jersey-quickstart-grizzly2/src/main/resources/archetype-resources/pom.xml:43]: Yes, agree. I externalized it with maven property

File: core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java [hunk 6]
```diff
@@ -147,27 +143,29 @@ public void testDefaultRxInvokerWithExecutorServiceProvider() throws ExecutionEx
...
     @Test
     public void testRxInvokerNotRegistered() {
-        Invocation.Builder request = target(CLIENT).request();
-        thrown.expect(IllegalStateException.class);
-        thrown.expectMessage(AllOf.allOf(
-                new StringContains("TestRxInvoker"),
-                new StringContains("not registered"),
-                new StringContains("RxInvokerProvider")));
-        request.rx(TestRxInvoker.class).get();
+        Assertions.assertThrows(IllegalStateException.class, () -> {
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java:161]: JUnit 5 has support-class [ExpectedExceptionSupport](https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/migrationsupport/rules/ExpectedExceptionSupport.html) for usage JUnit 4 Rule [ExpectedException](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html), but this support class also recomments to switch to assertThrows where possible ([source](https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/migrationsupport/rules/ExpectedExceptionSupport.html)):
```
However, you should rather switch to [Assertions.assertThrows(java.lang.Class<T>, org.junit.jupiter.api.function.Executable)](https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/api/Assertions.html#assertThrows(java.lang.Class,org.junit.jupiter.api.function.Executable)) for new code.
```
Of course I can restore usage of obsolete Rule,if you think it will better.

[the remark this answers is not in the document]
Comment (senivam) [reply core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java:161]: my idea actually was just to a bit extend of what you are offering:
```
final IllegalStateException exception = Assertions.assertThrows(IllegalStateException.class, () -> {
            Invocation.Builder request = target(CLIENT).request();
            request.rx(TestRxInvoker.class).get();
        });
      
        final String message = exception.getMessage();
        assertTrue(AllOf.allOf(
                new StringContains("TestRxInvoker"),
                new StringContains("not registered"),
                new StringContains("RxInvokerProvider"))
                .matches(message)
        );
```
or something like that. I did not check my code, so the test might be failing as is but this is just an example.

**Comment by bvfalcon** [comment]
> AccessControl access denied errors on CI Jenkins JDK 19
```
[ERROR] Errors: 
[ERROR]   ByteBufferInputStreamTest.testBlockingReadByteArray:408 » AccessControl access denied ("java.lang.RuntimePermission" "setContextClassLoader")
[ERROR]   ByteBufferInputStreamTest.testBlockingReadSingleByte:340 » AccessControl access denied ("java.lang.RuntimePermission" "setContextClassLoader")
```


> There is no Security Manager on JDK 19. 


Yes, Security manager is deprecated in Java 1 and removed in Java 19, but Security policies yet actual and need to be set. Additional policy and some maven build settings changes [help](https://github.com/eclipse-ee4j/jersey/pull/5123/commits/2fb8fcfec3bfeadfe80b576aa42e26e3fa0e0ef9).

File: core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (bvfalcon) [reply core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java]: Yes, aou are right. This check was lost during migration.
Restored!

**Review by senivam — CHANGES_REQUESTED** [review]
the review is still in progress, the following comments were found yesterday, so I'm posting them now in order not to delay the feedback.

File: core-common/src/test/resources/surefire-jdk17.policy [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2014, 2022 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply core-common/src/test/resources/surefire-jdk17.policy:2]: Yes, was copied from another file. Frankly, I am not an employee of Oracle, so not sure, that this &copy; will be correct. In other files there are Oracle copyright stand too and I modify existed &copy; but this file is new. I write here copyright with my name to avoid misunderstandings.

File: core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java [hunk 8]
```diff
@@ -147,27 +143,29 @@ public void testDefaultRxInvokerWithExecutorServiceProvider() throws ExecutionEx
...
 
     @Test
     public void testRxInvokerInvalid() {
-        Invocation.Builder request = target(CLIENT).request();
-        thrown.expect(IllegalArgumentException.class);
-        thrown.expectMessage(AllOf.allOf(new StringContains("null"), new StringContains("clazz")));
-        request.rx(null).get();
+        Assertions.assertThrows(IllegalArgumentException.class, () -> {
+            Invocation.Builder request = target(CLIENT).request();
+            // thrown.expectMessage(AllOf.allOf(new StringContains("null"), new StringContains("clazz")));
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply core-client/src/test/java/org/glassfish/jersey/client/ClientRxTest.java]: Outdated notice: was already changed in latest commits

File: examples/osgi-http-service/functional-test/pom.xml [hunk 9]
```diff
@@ -218,7 +219,7 @@
                                 <!--
                                 excluded due to javax.servlet-api collision -
                                 more than one version is required, so it is not possible to execute that test.
-                                (since pax-exam version 4.13.4)
+                                (since pax-exam version 4.13.5)
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply examples/osgi-http-service/functional-test/pom.xml:222]: Not so. javax.servlet-api collision [is not only reason](https://github.com/bvfalcon/jakarta-jersey/blob/junit-switch-4-to-5/pom.xml#L2191) to stay pax.exam on 4.13.4

File: test-framework/maven/custom-enforcer-rules/pom.xml [hunk 10]
```diff
@@ -34,11 +34,11 @@
         <dependency>
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-enforcer-plugin</artifactId>
-            <version>1.4.1</version>
+            <version>3.1.0</version>
             <exclusions>
                 <exclusion>
-                    <groupId>junit</groupId>
-                    <artifactId>junit</artifactId>
+                    <groupId>org.junit.jupiter</groupId>
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply test-framework/maven/custom-enforcer-rules/pom.xml:40]: Considering that in this level in pom.xml already defined junit-jupiter dependency, this exclusion can be safe removed

**Review by senivam — CHANGES_REQUESTED** [review]
For now great job, I have a few files left, my current notes are below. But anyway, I appreciate the work which is done. Thanks.

**Review by senivam — CHANGES_REQUESTED** [review]
I think I'm done with the review. My final comments are below. Thanks for the work done and please address my remarks.

File: tests/e2e-entity/src/test/java/org/glassfish/jersey/tests/e2e/json/JsonTest.java [hunk 11]
```diff
@@ -49,18 +49,18 @@
...
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 /**
  * Common functionality for JSON tests that are using multiple JSON providers (e.g. MOXy, Jackson, Jettison).
  *
  * @author Michal Gajdos
  */
-public abstract class AbstractJsonTest extends JerseyTest {
+public abstract class JsonTest extends JerseyTest {
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply tests/e2e-entity/src/test/java/org/glassfish/jersey/tests/e2e/json/JsonTest.java:60]: Partially this renaming [is dictated by these considerations](https://github.com/eclipse-ee4j/jersey/pull/5123#discussion_r983438348), but now this class is used as [```JsonTest jsonTest = new JsonTest(setupTest) {};```](https://github.com/bvfalcon/jakarta-jersey/blob/junit-switch-4-to-5/tests/e2e-entity/src/test/java/org/glassfish/jersey/tests/e2e/json/GenericTest.java#L42).
Formulation ```AbstractJsonTest jsonTest = new AbstractJsonTest(setupTest) {};``` seems much worse, than without "Abstract"-prefix ;)

File: tests/e2e-entity/src/test/java/org/glassfish/jersey/tests/e2e/entity/filtering/EntityFilteringServerTest.java [hunk 12]
```diff
@@ -29,20 +29,11 @@
...
 
-import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.junit.runners.Suite;
+import org.junit.jupiter.api.Test;
 
 /**
  * @author Michal Gajdos
  */
-@RunWith(Suite.class)
```
[the remark this answers is not in the document]
Comment (bvfalcon) [reply tests/e2e-entity/src/test/java/org/glassfish/jersey/tests/e2e/entity/filtering/EntityFilteringServerTest.java:39]: > JsonEntityFilteringServerTest

This Test was changed so much that it is impossible to use Suite annotations now.

**Comment by jansupol** [comment]
@bvfalcon 
You made a great deal of work, thank you. I'm finishing the review and might have a couple of comments.

Jersey Test Framework is not used just in Jersey. The test framework was modified to be working with both Junit 4 and Juniper, and it should remain as such. There are other frameworks built atop the Jersey test framework. 

First, there are public classes that should not be dropped from the test framework such as `ConcurrentParameterizedRunner`, `ConcurrentRunner`, etc.

**Comment by bvfalcon** [comment]
> there are public classes that should not be dropped from the test framework such as ConcurrentParameterizedRunner, ConcurrentRunner, etc.

With last commit these files have been restored. And also all changes squashed into one big commit

**Comment by jansupol** [comment]
@bvfalcon I made a PR to your branch with updates to your PR so that Jersey can be built and all the tests are executed.

**Review by jansupol — APPROVED** [review]
After running the tests, the test results are the same as with the previous Junit 4 and it runs flawlessly in my local env.

The TEST execution time seems to be a bit higher than with Junit 4, but it seems to be the price of adopting the currently developed version of Junit.

**Comment by jansupol** [comment]
@bvfalcon Thank you for this. Great job!
